types/priv_validator.go

```diff
@@ -107,6 +111,10 @@ func GenPrivValidator() *PrivValidator {
 }
 
 func LoadPrivValidator(filePath string) *PrivValidator {
+	return LoadPrivValidatorWithSigner(filePath, nil)
+}
+
+func LoadPrivValidatorWithSigner(filePath string, signer Signer) *PrivValidator {
```

Comment:
Why bother with this ? Why not just 

```
pv := LoadPrivValidator(filePath)
pv.SetSigner(signer) 
```
?

Reply:
@ebuchman That was the fix for [my comment](https://github.com/tendermint/tendermint/pull/637#commitcomment-24019236) above.

Reply:
But can't we just call the `SetSigner` method to override ?

Reply:
If you follow through the logic of `LoadPrivValidator`, you'll find it eventually operates on the private key using the default signer. The call to `setPubKeyAndAddress` calls `privVal.Signer.PubKey()`, which in turn calls `ds.priv.PubKey()`.

Reply:
Right of course. If the priv_key is empty, this won't work. Thanks.

Reply:
Ok, so what I would recommend is to change LoadPrivValidator to check if the privkey is empty. If it is, then it should return, and the caller can then call SetSigner. If it's not empty, then it can do 

```
    privVal.Signer = NewDefaultSigner(privVal.PrivKey)                                                                                       
    privVal.setPubKeyAndAddress()  
```

using the priv key from the file.

Does that make sense?

I'm just wary of making new exposed functions :)

Reply:
I had assumed developers would borrow your file format to store their own info. Hence privkey wouldn't be empty, but would be something specific to the crypto implementation (perhaps a PKCS#11 key label).

In case it affects your decision, I've recently decided to leap over this basic implement step and head straight for a new privValidator that runs inside an HSM (as we discussed). So this change we are discussing would be for the benefit of others taking a simpler approach to key management.

Reply:
It's just occurred to me (now that I've started coding in earnest), that to benefit from your existing file format, the signer would need to be created *after* the priv_validator.json file has been read. So these changes might need some additional thought. I'll code up something early next week and will share around.